src/core/backend_context.cc

```diff
@@ -731,26 +733,27 @@ BackendInputCollector::SetFixedSizeInputTensor(
     }
 
     // Direct copy without intermediate pinned memory.
-    bool cuda_used = false;
-    status = CopyBuffer(
+    worker_pool_.AddTask(
         request_input->Name(), src_memory_type, src_memory_type_id,
         tensor_memory_type, tensor_memory_type_id, src_byte_size, src_buffer,
         tensor_buffer + tensor_buffer_offset + input_offset, stream_,
-        &cuda_used);
-    cuda_copy |= cuda_used;
-
-    if (!status.IsOk()) {
-      LOG_STATUS_ERROR(
-          InferenceResponse::SendWithStatus(
-              std::move(*response), TRITONSERVER_RESPONSE_COMPLETE_FINAL,
-              status),
-          "error setting TensorFlow input tensor");
-      return cuda_copy;
-    }
+        &cuda_used[idx]);
 
     input_offset += src_byte_size;
   }
 
+  Status status = worker_pool_.ProcessQueue();
```

Comment:
If you re-purpose `ProcessQueue()` to be waiting for all tasks to be finished as suggested, you can defer the call to this as late as possible to reduce the number of "consume tasks -> wait for completion" cycles.